chore: Set up release-please

btmills opened the pull request:
This configures the [release-please action][1] for `generator-eslint`. As we merge changes using [Conventional Commits formatting][2], release-please will create and update a pending release PR. That PR will contain a changelog of any commits merged since the prior release. When we're ready to do the release, we'll merge that PR, and the workflow in this PR will update the changelog and publish the release to GitHub and npm.

Separate from the code in this PR, I had to take a few manual steps:

1. Create a new [npm granular access token][3].
   1. Name it ~~`release-please $package-name`~~ `release-please`.
   1. Choose a reasonable expiration. I went with 365 days.
   1. Under Packages and Scopes > select "Read and write" access > select ~~"Only select packages and scopes" > select the package that we're setting up.~~ "All packages".
   1. The token should not have Organization access.
   1. Generate the token.
1. Copy the generated token as a ~~[new repository secret][4]~~ [new organization secret](https://github.com/organizations/eslint/settings/secrets/actions) named `NPM_TOKEN`, and give access to the desired repository. If the token already exists, edit it to include access for the desired repository.
1. Edit the four Twitter API [organization access tokens](https://github.com/organizations/eslint/settings/secrets/actions) to grant access to the repository.
1. In the [repository's Actions settings][5], under "Workflow permissions", ensure "Allow GitHub Actions to create and approve pull requests" is checked.

[1]: https://github.com/google-github-actions/release-please-action
[2]: https://www.conventionalcommits.org/en/v1.0.0/
[3]: https://www.npmjs.com/settings/eslintbot/tokens/granular-access-tokens/new
[4]: https://github.com/eslint/generator-eslint/settings/secrets/actions/new
[5]: https://github.com/eslint/generator-eslint/settings/actions

btmills commented on .github/workflows/release-please.yml:
```diff
@@ -0,0 +1,32 @@
+        if: ${{ steps.release.outputs.release_created }}
+      - run: npm ci
+        if: ${{ steps.release.outputs.release_created }}
+      - run: npm test
```
Currently, we run one final round of tests as part of releases. I'm assuming we want to keep doing that. In theory, the full test matrix will have already passed on the release PR in order for us to merge it, so this single run is adding little value. It also introduces some risk: if the release-please step creates the release on GitHub but a flaky test fails in this step, the `npm publish` step won't run. Then if we re-run this workflow, release-please will see that the release was already created, skipping the rest of this workflow. We'd have to manually delete the release from GitHub before re-running.

nzakas replied:
I think it's okay to skip testing here.

btmills replied:
It's gone!

btmills commented on .github/workflows/release-please.yml:
```diff
@@ -0,0 +1,32 @@
+        if: ${{ steps.release.outputs.release_created }}
+      - run: npm test
+        if: ${{ steps.release.outputs.release_created }}
+      - run: npm publish
```
I didn't clean up the existing eslint-release dependency and package.json scripts in this PR. I'm happy to include that if you'd like, but I figured we'd like to try release-please first while retaining the option of releasing in Jenkins until we're sure we're ready to drop the Jenkins flow.

bmish replied:
That seems reasonable.

btmills commented on .github/workflows/release-please.yml:
```diff
@@ -0,0 +1,32 @@
+jobs:
+  release-please:
+    runs-on: ubuntu-latest
+    permissions:
```
This repository is currently [configured](https://github.com/eslint/generator-eslint/settings/actions) to give workflows write access by default. Specifying only the precise [`$GITHUB_TOKEN` permissions](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs) that release-please needs for this job reduces surface area and makes this workflow copy-pasteable to repositories that only grant read access by default.

btmills reviewed (commented):
GitHub Actions workflows are notoriously difficult to develop and debug, so I got all of this working in https://github.com/btmills/release-please-test first. Feel free to reference that to see what the workflow is like. For example, https://github.com/btmills/release-please-test/pull/15 is a release PR. I noted any manual configuration steps along the way and included them in the commit message and PR body for reference for future repositories. While I can't promise that this'll work on the first try, I'm hopeful that I ironed out all the kinks over in that test repository.

bmish reviewed (approved):
This looks useful!

nzakas reviewed (commented):
It might be better to create the npm token secret at the organization level so it can be reused by all repos.

You can set the secret to only be available to this repo to start.

I just foresee a time where we need to update 20 tokens otherwise.

nzakas commented:
By the way, we could also tweet out the release, as I do in my repos:
https://github.com/humanwhocodes/async-event-emitter/blob/e63c3b054c2fcdf422fc2f75ff1053801589d7f7/.github/workflows/release-please.yml#L33-L39

We already have our Twitter credentials as secrets to tweet out RFC updates, and at least so far, these automated tweets are still going out.

btmills commented:
1. I moved `NPM_TOKEN` to the [organization level](https://github.com/organizations/eslint/settings/secrets/actions) and gave `generator-eslint` access.
2. I added the tweet action. That's the only part of this that I haven't tested on my test repository, but I'm hoping the rfcs repository example is sufficient to get this one right on the first try 🤞

nzakas reviewed (approved):
LGTM!
